### PR TITLE
gh-95337: update TypeVarTuple example

### PR DESCRIPTION
**Pull request by adriangb**
Notice that "T" is only used once in the function signature, which Pylance will highlight as an error.
In this case, Any is just as meaningful as T (and I would argue semantically better, we don't care what the first element is)
All of the examples (checked with reveal_type) are still valid.

* Issue: gh-95337

### CODE REVIEW

**Comment by gvanrossum** [comment]
@mrahtz @pradeep90 Do you agree?

**Review by pradeep90 — CHANGES_REQUESTED** [review]
> In this case, Any is just as meaningful as T (and I would argue semantically better, we don't care what the first element is)

I'm against removing the `T`. It's there to demonstrate how tuples are split into `T` and `*Ts` and what each variable is bound to. Otherwise, there's nothing in the docs to show this key behavior. 

Given that this example is for explanation, I don't think we should change it just to avoid Pylance lint errors.

> Notice that "T" is only used once in the function signature, which Pylance will highlight as an error.

If you feel *really* strongly about the lint error, you could change it to `split_first_and_rest(xs: tuple[T, *Ts]) -> tuple[T, tuple[*Ts]]: ...`. But that complicates the introduction to a concept that can already be hard to understand. I think the example is fine as it is.

**Comment by adriangb** [comment]
> Given that this example is for explanation, I don't think we should change it just to avoid Pylance lint errors.

That's fair, I wasn't trying to pose that as a substantial point, that's just how I noticed it.

> I'm against removing the `T`. It's there to demonstrate how tuples are split into `T` and `*Ts` and what each variable is bound to. Otherwise, there's nothing in the docs to show this key behavior.

Then I think we should change the example to fully demonstrate that behavior (i.e. the more complicated version you just proposed or similar). As is, that behavior is encoded in comments but not the code. Personally I think `Any` demonstrates the behavior just fine and keeps the example simple (simpler in fact because there's one less type variable to grok).

**Comment by mrahtz** [comment]
Sorry for the slow reply - busy week.

I agree with:
> It's there to demonstrate how tuples are split into T and *Ts and what each variable is bound to. Otherwise, there's nothing in the docs to show this key behavior.

To put it another way, I think it's important to emphasise that TypeVarTuples aren't anything 'special' - that the mental model is "TypeVarTuples just unpack into something that looks like a bunch of TypeVars, so it can be used with other regular TypeVars" - as opposed to "TypeVarTuple is this special thing that needs to be used on its own and I have to arbitrarily star", which is the nuance I think a reader _might_ get from
```python
def remove_first_element(tup: tuple[Any, *Ts]) -> tuple[*Ts]:
    return tup[1:]
```
(Ok, yes, this example does at least mix a TypeVarTuple with an Any - but that might not be enough to give the impression that TypeVarTuples can also be mixed with regular TypeVars.)

Having said that, I also agree with:
> Then I think we should change the example to fully demonstrate that behavior (i.e. the more complicated version you just proposed or similar).

More than just fixing Pylance lint, I do think it's important our examples implicitly instruct the user in what kinds of things are valid to do. Since (iiuc?) most type checkers do enforce this thing that it's not valid to have an unreferenced type variable, I think it would be good to avoid giving the user the impression that this might be possible.

> But that complicates the introduction to a concept that can already be hard to understand. I think the example is fine as it is.

It does make it more complicated, but only a little bit - iiuc:
```python
 def move_first_element_to_last(tup: tuple[T, *Ts]) -> tuple[*Ts, T]:
   return (*tup[1:], tup[0])
```
If the reader understands the splitting of `tuple[T, *Ts]` in the argument type, I think they also should be fine with it in the return type - also there's also the nice parallel of the star in the method body to the star in the type signature, emphasising that star in the type signature has the same meaning as regular unpacking.

Having said that, I don't feel strongly about this, so Pradeep, if you still have hesitations, I trust your judgement, and we can leave it as it is.

**Comment by adriangb** [comment]
Thank you for the feedback folks. It sounds like the consensus is that if there was going to be a change made it would be in the direction of the more complex example that uses the type var in the return type, so I'll at least change this PR to propose that instead of using Any.

**Comment by JelleZijlstra** [comment]
I like Matthew's idea of having the example move the first tuple element to the end: it's a fairly realistic example and yet not too complicated.

**Comment by adriangb** [comment]
That's a great example @mrahtz , I had missed it. I like that it also uses `tuple[*Ts, T]` which shows that the TypeVarTuple can be unpacked anywhere and not just the last/remainder.

For completeness, I added the invalid case of `move_first_element_to_last(tup=())`

File: Doc/library/typing.rst [hunk 1]
```diff
@@ -1305,20 +1305,25 @@ These are not used in annotations. They are building blocks for creating generic
...
-        remove_first_element(tup=(1, 'spam'))
+        # Return value is ('spam', 1), which has type tuple[str, int]
+        move_first_element_to_last(tup=(1, 'spam'))
 
         # T is bound to int, Ts is bound to (str, float)
-        # Return value is ('spam', 3.0), which has type tuple[str, float]
-        remove_first_element(tup=(1, 'spam', 3.0))
+        # Return value is ('spam', 3.0, 1), which has type tuple[str, float, int]
+        move_first_element_to_last(tup=(1, 'spam', 3.0))
+
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]:
```suggestion

```

File: Doc/library/typing.rst [hunk 2]
```diff
@@ -1305,20 +1305,24 @@ These are not used in annotations. They are building blocks for creating generic
...
-        remove_first_element(tup=(1, 'spam'))
+        # Return value is ('spam', 1), which has type tuple[str, int]
+        move_first_element_to_last(tup=(1, 'spam'))
 
         # T is bound to int, Ts is bound to (str, float)
-        # Return value is ('spam', 3.0), which has type tuple[str, float]
-        remove_first_element(tup=(1, 'spam', 3.0))
+        # Return value is ('spam', 3.0, 1), which has type tuple[str, float, int]
+        move_first_element_to_last(tup=(1, 'spam', 3.0))
+        # This fails to type check (and at runtime)
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]:
```suggestion

        # This fails to type check (and fails at runtime)
```

**Review by AlexWaygood — APPROVED** [review]
I like this PR!

File: Doc/library/typing.rst [hunk 3]
```diff
@@ -1305,20 +1305,25 @@ These are not used in annotations. They are building blocks for creating generic
...
 
-        def remove_first_element(tup: tuple[T, *Ts]) -> tuple[*Ts]:
-            return tup[1:]
+        def move_first_element_to_last(tup: tuple[T, *Ts]) -> tuple[*Ts, T]:
+            return (*tup[1:], tup[0])
 
         # T is bound to int, Ts is bound to ()
-        # Return value is (), which has type tuple[()]
-        remove_first_element(tup=(1,))
+        # Return value is (1, ), which has type tuple[int]
```
Comment (mrahtz) [inline Doc/library/typing.rst]: Nit: I think PEP 8 says there shouldn't be a space after the comma here: `(1,)`

**Review by mrahtz — CHANGES_REQUESTED** [review]
Lgtm. Thanks, Adrian!

File: Doc/library/typing.rst (hunk 3 above)
Comment (adriangb) [reply Doc/library/typing.rst]: done ✅

**Review by JelleZijlstra — APPROVED** [review]

**Review by JelleZijlstra — APPROVED** [review]

**Comment by AlexWaygood** [comment]
Oops, my bad, no TypeVarTuples in 3.10

**Pull request merged**